File: lib/api/models/configuration.dart
// lib/models/configuration.dart

class Configuration {
  final int? configId;
  final String apiKey;
  final String modelVersion;
  final String endpointUrl;
  final int requestTimeout;
  final int maxTokens;
  final double temperature;
  final bool isActive;
  final int? updatedByUserId;
  final DateTime? updatedAt;

  // Related user
  final Map<String, dynamic>? user;

  Configuration({
    this.configId,
    required this.apiKey,
    required this.modelVersion,
    required this.endpointUrl,
    this.requestTimeout = 30,
    this.maxTokens = 2048,
    this.temperature = 0.7,
    this.isActive = false,
    this.updatedByUserId,
    this.updatedAt,
    this.user,
  });

  factory Configuration.fromJson(Map<String, dynamic> json) {
    return Configuration(
      configId: json['config_id'],
      apiKey: json['api_key'],
      modelVersion: json['model_version'],
      endpointUrl: json['endpoint_url'],
      requestTimeout: json['request_timeout'] ?? 30,
      maxTokens: json['max_tokens'] ?? 2048,
      temperature: (json['temperature'] != null)
          ? (json['temperature'] as num).toDouble()
          : 0.7,
      isActive: json['is_active'] ?? false,
      updatedByUserId: json['updated_by_user_id'],
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : null,
      user:
          json['user'] != null ? Map<String, dynamic>.from(json['user']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (configId != null) 'config_id': configId,
      'api_key': apiKey,
      'model_version': modelVersion,
      'endpoint_url': endpointUrl,
      'request_timeout': requestTimeout,
      'max_tokens': maxTokens,
      'temperature': temperature,
      'is_active': isActive,
      if (updatedByUserId != null) 'updated_by_user_id': updatedByUserId,
      if (updatedAt != null) 'updated_at': updatedAt!.toIso8601String(),
      if (user != null) 'user': user,
    };
  }

  Configuration copyWith({
    int? configId,
    String? apiKey,
    String? modelVersion,
    String? endpointUrl,
    int? requestTimeout,
    int? maxTokens,
    double? temperature,
    bool? isActive,
    int? updatedByUserId,
    DateTime? updatedAt,
    Map<String, dynamic>? user,
  }) {
    return Configuration(
      configId: configId ?? this.configId,
      apiKey: apiKey ?? this.apiKey,
      modelVersion: modelVersion ?? this.modelVersion,
      endpointUrl: endpointUrl ?? this.endpointUrl,
      requestTimeout: requestTimeout ?? this.requestTimeout,
      maxTokens: maxTokens ?? this.maxTokens,
      temperature: temperature ?? this.temperature,
      isActive: isActive ?? this.isActive,
      updatedByUserId: updatedByUserId ?? this.updatedByUserId,
      updatedAt: updatedAt ?? this.updatedAt,
      user: user ?? this.user,
    );
  }
}

File: final_project/models/configuration.py
from datetime import datetime

from sqlalchemy import (
    Column, Integer, String, Text, Boolean, DateTime,
    Numeric, ForeignKey, Index, event
)
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship

from ..db import Base

class ConfigurationDB(Base):
    __tablename__ = "configuration"

    config_id = Column(
        Integer,
        primary_key=True,
        autoincrement=True
    )  # Primary key
    api_key = Column(
        Text,
        nullable=False
    )  # AI API key
    model_version = Column(
        String(100),
        nullable=False
    )  # AI model version
    endpoint_url = Column(
        Text,
        nullable=False
    )  # API endpoint URL
    request_timeout = Column(
        Integer,
        default=30,
        nullable=False
    )  # Request timeout (seconds)
    max_tokens = Column(
        Integer,
        default=2048,
        nullable=False
    )  # Max tokens per request
    temperature = Column(
        Numeric(3, 2),
        default=0.7,
        nullable=False
    )  # Sampling temperature
    is_active = Column(
        Boolean,
        default=False,
        nullable=False
    )  # Whether this config is active

    # Reference to the user who last updated this configuration
    updated_by_user_id = Column(
        Integer,
        ForeignKey("users.user_id", ondelete="SET NULL"),
        nullable=True
    )
    updated_at = Column(
        DateTime,
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False
    )  # Last updated timestamp

    __table_args__ = (
        Index("idx_configuration_is_active", "is_active"),
    )

    # Relationship to UserDB
    user = relationship(
        "UserDB",
        back_populates="configurations",
        foreign_keys=[updated_by_user_id]
    )

# Event listener to update 'updated_at' timestamp before updates
@event.listens_for(ConfigurationDB, 'before_update')
def _update_timestamp(mapper, connection, target):
    target.updated_at = datetime.utcnow()